src/prerender.ts

```diff
@@ -0,0 +1,354 @@
+					prerenderEntry = bundle[output];
+				}
+			}
+
+			let head: Head = { lang: "", title: "", elements: new Set() };
+
+			if (!prerenderEntry) {
+				this.error("Cannot detect module with `prerender` export");
+			}
+
+			const m = await import(
+				`file://${path.join(tmpDir, path.basename(prerenderEntry.fileName))}`
+			);
+			const prerender = m.prerender;
+
+			if (typeof prerender !== "function") {
+				this.error("Detected `prerender` export, but it is not a function");
+			}
+
+			// We start by pre-rendering the home page.
+			// Links discovered during pre-rendering get pushed into the list of routes.
+			const seen = new Set(["/", ...additionalPrerenderRoutes!]);
+
+			let routes: PrerenderedRoute[] = [...seen].map(link => ({ url: link }));
+
+			for (const route of routes) {
+				if (!route.url) continue;
+
+				const outDir = route.url.replace(/(^\/|\/$)/g, "");
+				const assetName = path.join(
+					outDir,
+					outDir.endsWith(".html") ? "" : "index.html",
+				);
+
+				// Update `location` to current URL so routers can use things like `location.pathname`
+				const u = new URL(route.url, "http://localhost");
+				for (let i in u) {
+					try {
+						// @ts-ignore
+						globalThis.location[i] = String(u[i]);
+					} catch {}
+				}
+
+				const result = await prerender({ ssr: true, url: route.url, route });
+				if (result == null) continue;
+
+				// Reset HTML doc & head data
+				htmlDoc = htmlParse(tpl);
+				head = { lang: "", title: "", elements: new Set() };
+
+				// Add any discovered links to the list of routes to pre-render:
+				if (result.links) {
+					for (let url of result.links) {
+						const parsed = new URL(url, "http://localhost");
+						url = parsed.pathname;
+						// ignore external links and ones we've already picked up
+						if (seen.has(url) || parsed.origin !== "http://localhost") continue;
+						seen.add(url);
+						routes.push({ url, _discoveredBy: route });
+					}
+				}
+
+				let body;
+				if (result && typeof result === "object") {
+					if (result.html) body = result.html;
+					if (result.head) {
+						head = result.head;
+					}
+				} else {
+					body = result;
+				}
+
+				const htmlHead = htmlDoc.querySelector("head");
+				if (htmlHead) {
+					if (head.title) {
+						const htmlTitle = htmlHead.querySelector("title");
+						htmlTitle
+							? htmlTitle.set_content(enc(head.title))
+							: htmlHead.insertAdjacentHTML(
+									"afterbegin",
+									`<title>${enc(head.title)}</title>`,
+							  );
+					}
+
+					if (head.lang) {
+						htmlDoc.querySelector("html")!.setAttribute("lang", enc(head.lang));
+					}
+
+					if (head.elements) {
+						// Inject HTML links at the end of <head> for any stylesheets injected during rendering of the page:
+						htmlHead.insertAdjacentHTML(
+							"beforeend",
+							Array.from(
+								new Set(Array.from(head.elements).map(serializeElement)),
+							).join(""),
```

Comment:
For readability, this allows folks to validate the `head` tags in the generated html files

```suggestion
							).join("\n"),
```

Reply:
Good shout